pkg/murmur3/murmur3.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package murmur3
+
+import (
+	"unsafe"
+)
+
+// Hash128 calculates a 128 bits hash for the given data. It returns different
+// results when running on big-endian and little-endian machines.
+func Hash128(data []byte, seed uint32) (uint64, uint64) {
+	nblocks := len(data) / 16
+
+	h1 := uint64(seed)
+	h2 := uint64(seed)
+
+	c1 := uint64(0x87c37b91114253d5)
+	c2 := uint64(0x4cf5ad432745937f)
+
+	for i := 0; i < nblocks; i++ {
+		tmp := (*[2]uint64)(unsafe.Pointer(&data[i*16]))
+		k1 := tmp[0]
+		k2 := tmp[1]
+
+		k1 *= c1
+		k1 = rotl64(k1, 31)
+		k1 *= c2
+		h1 ^= k1
+		h1 = rotl64(h1, 27)
+		h1 += h2
+		h1 = h1*5 + 0x52dce729
+		k2 *= c2
+		k2 = rotl64(k2, 33)
+		k2 *= c1
+		h2 ^= k2
+		h2 = rotl64(h2, 31)
+		h2 += h1
+		h2 = h2*5 + 0x38495ab5
+	}
+
+	k1 := uint64(0)
+	k2 := uint64(0)
+
+	tail := data[nblocks*16:]
+
+	switch len(tail) & 15 {
```

Comment:
It's not entirely clear why this code unfolded this way, e.g. if there a performance benefit or something, it would be good if there was a comment to clarify :)

Reply:
I can see that it's based on C++ code, maybe it could be optimised for readability, or at least just add a comment above to say that it's C++ code that was converted to Go and unchanged for the rest.

Reply:
I wanted to keep the implementation as close as possible to the reference implementation in cpp, so that it'd be easier to compare both of them. IMHO, adding comments would hinder that.

Reply:
Just a comment at the top of the file maybe?

Reply:
Maybe having unit tests would help us to adhere to the spec? Does the original C++ code have some tests?

Reply:
@errordeveloper Please see https://github.com/cilium/cilium/pull/13501/files#diff-5106668213bdbc42f6220cefac2488e57a05dd964f310533b212f0650d63120cR34 (pkg/murmur3/murmur3_tests.go) :-)